arch/lkl/kernel/cpu.c

```diff
@@ -150,6 +156,53 @@ void lkl_cpu_put(void)
 	lkl_ops->mutex_unlock(cpu.lock);
 }
 
+#ifdef DEBUG
+
+/*
+ *	Debug tool. Essentially allows for assert(cpuLockTaken);
+ */
+
+static int lkl_check_cpu_owner()
+{
+	int result;
+	lkl_ops->mutex_lock(cpu.lock);
+	lkl_thread_t self = lkl_ops->thread_self();
+	lkl_thread_t owner = cpu.owner;
+	if (!cpu.count || !owner ||
```

Comment:
According to comments at the top of the file /* no of times the current thread got the CPU */

I don't like this lkl_cpu locking scheme. It is complicated and hard to understand. The logic in this function (lkl_check_cpu_owner) is modelled on various other, pre-existing, checks such as in lkl_cpu_put which might otherwise go off. The idea is to allow code that has locking preconditions to check and always fail rather than often get away with it and have intermittent test failures. Figuring out the call tree above is hard and doesn't prevent later changes allowing calls without the correct lock state.

In particular we need the lock to be taken exactly once so that we are safe up to the point the user code is handled but can release the lock  over the user code.

Reply:
Thank you for the explanation.